src/libraries/System.Private.CoreLib/src/System/IO/BufferedStream.cs

```diff
@@ -1210,38 +1231,23 @@ public override long Seek(long offset, SeekOrigin origin)
                 return _stream.Seek(offset, origin);
             }
 
-            // The buffer is either empty or we have a buffered read.
-
-            if (_readLen - _readPos > 0 && origin == SeekOrigin.Current)
+            long newPos;
+            switch (origin)
             {
-                // If we have bytes in the read buffer, adjust the seek offset to account for the resulting difference
-                // between this stream's position and the underlying stream's position.
-                offset -= (_readLen - _readPos);
-            }
-
-            long oldPos = Position;
-            Debug.Assert(oldPos == _stream.Position + (_readPos - _readLen));
-
-            long newPos = _stream.Seek(offset, origin);
-
-            // If the seek destination is still within the data currently in the buffer, we want to keep the buffer data and continue using it.
-            // Otherwise we will throw away the buffer. This can only happen on read, as we flushed write data above.
-
-            // The offset of the new/updated seek pointer within _buffer:
-            _readPos = (int)(newPos - (oldPos - _readPos));
```

Comment:
I was not aware of the fact that `Seek` was capable of reusing the buffer while `Position` setter was not. Since we are very close to release of .NET 6 and I don't feel comfortable introducing any non trivial changes to the buffering logic (I am speaking from experience, as I've blocked preview 4 release with a bug in buffering logic myself: #51151) I would prefer to reuse the existing `Seek` logic by simply calling `Seek` from `Position` setter.

@lateapexearlyspeed please don't  get me wrong. I am not saying that your refactor and logic changes have a bug, I just don't feel 100% comfortable merging it now.

Reply:
> I would prefer to reuse the existing Seek logic by simply calling Seek from Position setter

set_Position already calls Seek; it just first zeros out some state... it's not clear to me why it does that, but it would seem the right answer here would just be to delete code from the setter:
```diff
            set
            {
                if (value < 0)
                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.value, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
 
-                EnsureNotClosed();
-                EnsureCanSeek();
- 
-                if (_writePos > 0)
-                    FlushWrite();
- 
-                _readPos = 0;
-                _readLen = 0;
                _stream!.Seek(value, SeekOrigin.Begin);
            }
```
Is that sufficient to address the issue?

Reply:
> set_Position already calls Seek

it calls `_stream!.Seek`, while it should call `this.Seek(value, SeekOrigin.Begin)` which contains the logic that we care about

Reply:
> it calls _stream!.Seek, while it should call this.Seek(value, SeekOrigin.Begin) which contains the logic that we care about

Ah, I missed the `_stream.` part. Then yeah, that last line should change as well.
```diff
            set
            {
                if (value < 0)
                    ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.value, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
 
-                EnsureNotClosed();
-                EnsureCanSeek();
- 
-                if (_writePos > 0)
-                    FlushWrite();
- 
-                _readPos = 0;
-                _readLen = 0;
-                _stream!.Seek(value, SeekOrigin.Begin);
+               Seek(value, SeekOrigin.Begin);
            }
```